# Analytics Decision Making
## Individual Homework 1
### Shaolong (Fred) Xue

### **Exercise 1**

#### **Question 1**

**Optimization Model for Plant A**

Decision Variables:

* SA = units of Standard product from Plant A
* DA = units of Deluxe product from Plant A

Objective Function:

* Max(profit) = 10SA + 15DA

Constraints:

* Grinding Capacity: The total grinding hours used by Standard and Deluxe products cannot exceed the grinding capacity of the plant.
  * 4SA + 2DA ≤ 80

* Polishing Capacity: The total polishing hours used by Standard and Deluxe products cannot exceed the polishing capacity of the plant.
  * 2SA + 5DA ≤ 60

* Raw Material Usage: The raw material used by both products cannot exceed the allocation for the plant.
  * 4*(SA + DA) ≤ 75

**Optimization Model for Plant B**

Decision Variables:

* SB = units of Standard product from Plant B
* DB = units of Deluxe product from Plant B

Objective Function:

* Max(profit) = 10SB + 15DB

Constraints:

* Grinding Capacity: The total grinding hours used by Standard and Deluxe products cannot exceed the grinding capacity of the plant.
  * 5SB + 3DB ≤ 60

* Polishing Capacity: The total polishing hours used by Standard and Deluxe products cannot exceed the polishing capacity of the plant.
  * 5SB + 6DB ≤ 75

* Raw Material Usage: The raw material used by both products cannot exceed the allocation for the plant.
  * 4*(SB + DB) ≤ 45

#### **Question 2**

In [6]:
import numpy as np
import cvxpy as cp

In [7]:
# Define the decision variables
SA = cp.Variable(integer=True)
DA = cp.Variable(integer=True)
SB = cp.Variable(integer=True)
DB = cp.Variable(integer=True)

# Define the constraints
constraints_A = [
    4*SA + 2*DA <= 80,    # Grinding capacity for Plant A
    2*SA + 5*DA <= 60,    # Polishing capacity for Plant A
    4*(SA + DA) <= 75,    # Raw material capacity for Plant A
    SA >= 0,
    DA >= 0
]

constraints_B = [
    5*SB + 3*DB <= 60,    # Grinding capacity for Plant B
    5*SB + 6*DB <= 75,    # Polishing capacity for Plant B
    4*(SB + DB) <= 45,    # Raw material capacity for Plant B
    SB >= 0,
    DB >= 0
]

# Define the objective function
objective_A = cp.Maximize(10*SA + 15*DA)
objective_B = cp.Maximize(10*SB + 15*DB)

# Define the problem
probA = cp.Problem(objective_A, constraints_A)
probB = cp.Problem(objective_B, constraints_B)

# Solve the problem
probA.solve(solver=cp.GLPK_MI)
probB.solve(solver=cp.GLPK_MI)

# Print the result
print("Plant A: SA =", SA.value, ", DA =", DA.value, ", Total Profit = $", probA.value)
print("Plant B: SB =", SB.value, ", DB =", DB.value, ", Total Profit = $", probB.value)

Plant A: SA = 10.0 , DA = 8.0 , Total Profit = $ 220.0
Plant B: SB = 0.0 , DB = 11.0 , Total Profit = $ 165.0


#### **Question 3**

In [8]:
# Define the decision variables
SA = cp.Variable(integer=True)
DA = cp.Variable(integer=True)
SB = cp.Variable(integer=True)
DB = cp.Variable(integer=True)

# Define the constraints
constraints = [
    4*SA + 2*DA <= 80,    # Grinding capacity for Plant A
    2*SA + 5*DA <= 60,    # Polishing capacity for Plant A
    5*SB + 3*DB <= 60,    # Grinding capacity for Plant B
    5*SB + 6*DB <= 75,    # Polishing capacity for Plant B
    4*(SA + DA + SB + DB) <= 120, # Raw material for both plants
    SA >= 0,
    DA >= 0,
    SB >= 0,
    DB >= 0
]

# Define the objective function
objective = cp.Maximize(10*SA + 15*DA + 10*SB + 15*DB)

# Define the problem
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve(solver=cp.GLPK_MI)

# Print the result
print("Joint Optimization: SA =", SA.value, ", DA =", DA.value, ", SB =", SB.value, ", DB =", DB.value, ", Total Profit = $", problem.value)

Joint Optimization: SA = 10.0 , DA = 8.0 , SB = 0.0 , DB = 12.0 , Total Profit = $ 400.0


#### **Question 4**

In Question 2, we optimized the operations of Plant A and Plant B separately. The results were:

* Plant A: SA = 10.0 , DA = 8.0 , Total Profit = 220.0
* Plant B: SB = 0.0 , DB = 11.0 , Total Profit = 165.0

In Question 3, we jointly optimized the operations of Plant A and Plant B, treating them as a single system. The results were:

* Joint Optimization: SA = 10.0 , DA = 8.0 , SB = 0.0 , DB = 12.0 , Total Profit = 400.0

Overall all, production in Plant A was the same, but Plant B has one unit increase in producing deluxe products. This drives the total profit up from 385 to 400. 

The reason for this increase is due to the relaxation of raw material constraint in individual plants. The total amount of raw materials available across both plants was set to 120 kg, but there was no fixed allocation per plant.

#### **Question 5**

Based on the analysis, I recommend the CEO to shift towards a joint optimization approach for its operations. As we've showned you previously, individual operation at Plant A and Plant B results in a waste of raw material capacity and unrealized gains. If we treat the two plants as a single system and dynamically allocate raw materials based on demand and operational capacities, we can increase the total profit from \$385 to $400 per week. In particular, it would be beneficial to increase the production of the deluxe product at Plant B by utilizing the extra raw materials available. The production plan for Plant A can remain the same. This joint optimization approach provides flexibility and allows us to better manage our resources, resulting in more effective operations and higher profits. It's also crucial that we continually reassess these models, as changes in market conditions, operational capacities, or raw material availability can shift optimal production strategies.

### **Exercise 2**

#### **Question 1**

**Positive Effects on Profitability**:

* Price of the Inn: The coefficient is 5.86, which indicates that as the price of the Inn increases, the profitability also increases, assuming all other factors remain constant.
* College Students in the Area: The coefficient is 1.75, suggesting that an increase in the number of college students in the area would also lead to an increase in profitability, with all else being equal.

**Negative Effects on Profitability**:

* State Population per Inn: The coefficient is -5.41, implying that an increase in the state population per Inn reduces profitability, keeping all other variables constant.
* Square Root of the Median Income in the area: The coefficient is -3.09, suggesting that as the square root of the median income in the area increases, profitability decreases, assuming all other factors remain constant.

**Intuitive, these effects make sense**:
* The positive effect of 'Price of the Inn' is straightforward. The higher the price of staying at the Inn, the higher the revenue and thus, potentially, the higher the profitability, assuming costs remain constant.
* 'College Students in the Area' may frequent the Inn, boosting its profits. They may also attract more visitors, or indicate seasonal activities such as commencement, homecoming, etc., further enhancing profitability.
*The negative effect of 'State Population per Inn' might be due to competition. If the ratio of the state population per Inn is high, it may indicate many inns serving the same population, which could reduce the profitability of each inn.
* The negative effect of 'Square Root of the Median Income in the area' is a bit counter-intuitive as one might expect that higher income could lead to more spending in hotels. However, it might be the case that areas with higher median income have residents that travel more, and hence, use the local inns less. It could also be that these areas have more expensive real estate and operating costs, reducing profitability.

#### **Question 2**

I have calculated profitability on Excel. This updated file is also submitted with this assignment. 

Based on the calculation, 
* the hotel with the highest profitability is Hotel 2, located at Fresno, California. 
* the hotel with the lowest profitability is Hotel 8, located at Los Angeles, California. 

I am loading this file to Google Colab for later questions.

In [22]:
import pandas as pd

# File path for "LQM.xlsx" on my Google Drive
path = "/LQM.xlsx"

In [23]:
LQM = pd.read_excel(path, header=2)

/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


In [24]:
LQM.head(5)

Hotel                Location     Price  Price (normalized)  \
0      2      Fresno, California  10000000            1.699076   
1      6  Long Beach, California   8950000            1.402123   
2      1      Eureka, California   2925000           -0.301823   
3      3      Fresno, California   3750000           -0.068503   
4      4      Fresno, California   3500000           -0.139206   

   Square Root of Median Income (normalized)  \
0                                  -0.408199   
1                                   0.657845   
2                                  -0.812780   
3                                  -0.408199   
4                                  -0.408199   

   College Students in Area (normalized)  \
0                               0.311669   
1                               0.483711   
2                              -0.536413   
3                               0.311669   
4                               0.311669   

   State Population Per Inn (normalized)  Profitability  
0                              -0.474279      53.379192  
1                              -0.557278      49.095069  
2                              -0.995987      44.242369  
3                              -0.474279      43.021179  
4                              -0.474279      42.606858

#### **Question 3**

If we implement the "greedy" approach: 

**Part a:**

We would just purchase Hotel 2 (price = $10M) and run out of our budget. 

**Part b:**

Our total predicted profitability would just be the profitiability of Hotel 2, which is 53.38.

**Part c:**
* If we're just trying to maximize **total** predicted prodicted profitability, the "greedy" approach may not be the best. Because there may be a combination of less expensive hotels that provide a higher total profitability than Hotel 2 alone. 
* If we are trying to maximize **average** predicted profitability, the "greedy" approach could work well, especially when the budget is limited. Because buying the most profitable hotel would simply maximize the average profitability. 
* If our budget is /$20M, the same holds for maximizing **total** predicted profitability. In terms of **average** predicted profitability, that doesn't help us because sticking to buying the single most profitable hotel will always maximize the average. Adding any hotels with lower profitability than Hotel 2 will lower the average predicted profitability.

#### **Question 4**

**Part a:**

* Decision Variables:
Let's define binary decision variables x[i] for i in [1, 2, ..., 16]. x[i] is equal to 1 if the i-th hotel is selected and 0 otherwise.

* Objective Function:
The objective is to maximize the total profitability. So, the objective function would be the sum of the profitability of each hotel multiplied by its associated decision variable.

  * Maximize ∑ (Profitability[i] * x[i]) for all i

* Constraints:
The total cost of the selected hotels must be less than or equal to the budget of $10 million.

  * ∑ (Price[i] * x[i]) for all i <= 10,000,000


**Part b:**

In [27]:
# Define the decision variables
x = cp.Variable(len(LQM), boolean=True)

# Define the objective function
objective = cp.Maximize(cp.sum(cp.multiply(LQM['Profitability'], x)))

# Define the budget constraint
budget_constraint = cp.sum(cp.multiply(LQM['Price'], x)) <= 10000000

# Define the problem
problem = cp.Problem(objective, [budget_constraint])

# Solve the problem
problem.solve()

# Get the optimal solution
optimal_decision = x.value
optimal_objective = problem.value

# Print the optimal decision variables and objective function value
print("Optimal Decision Variables:")
for i, decision in enumerate(optimal_decision):
    if decision > 0.5:
        print("Hotel", LQM['Hotel'].iloc[i])

print("Optimal Objective Function Value:", optimal_objective)

Optimal Decision Variables:
Hotel 12
Hotel 13
Hotel 10
Hotel 15
Hotel 11
Hotel 16
Hotel 5
Optimal Objective Function Value: 269.9246813771899


**Part c:**

This solution makes sense intuitively. With this optimized solution, we would acquire 7 hotels, instead of just one hotel in the "greedy" solution. Our total predicted profitability, given the /$10M budget, would be 269.92. This is a significant increase from just 53.38 in the "greedy" solution. 

#### **Question 5**

**Part a:**

We need to add a constraint that sums up the decision variables x for each unique city and ensures that the sum is less than or equal to 2. This effectively limits the number of hotels purchased in any city to at most 2.

Intuitively, adding this constraint will probably result in a smaller optimal objective function value compared to before, because we are now constrained to diversify across as many cities as possible. And it might forces us to select hotels with lower profitability but in different cities to satisify the constraint. That is the trade-off between diversifying the LQM brand to different locations and achieving highest total profitability. 

**Part b & c:**

In [31]:
# Define the decision variables
x = cp.Variable(len(LQM), boolean=True)

# Define the objective function
objective = cp.Maximize(cp.sum(cp.multiply(LQM['Profitability'], x)))

# Define the budget constraint
budget_constraint = cp.sum(cp.multiply(LQM['Price'], x)) <= 10000000

# Define the city constraint
city_constraint = []
for location in LQM['Location'].unique():
    city_constraint.append(cp.sum(x[LQM['Location'] == location]) <= 2)

# Define the problem
problem = cp.Problem(objective, [budget_constraint] + city_constraint)

# Solve the problem
problem.solve()

# Get the optimal solution
optimal_decision = x.value
optimal_objective = problem.value

# Print the optimal solution
print("Optimal Decision:")
for i in range(len(optimal_decision)):
    print(f"Hotel {i+1}: {optimal_decision[i]}")

# Print the optimal objective function value
print("Optimal Objective Value:", optimal_objective)

Optimal Decision:
Hotel 1: 0.0
Hotel 2: 0.0
Hotel 3: 1.0
Hotel 4: 0.0
Hotel 5: 0.0
Hotel 6: 0.0
Hotel 7: 0.0
Hotel 8: 0.0
Hotel 9: 1.0
Hotel 10: 0.0
Hotel 11: 1.0
Hotel 12: 0.0
Hotel 13: 1.0
Hotel 14: 0.0
Hotel 15: 1.0
Hotel 16: 1.0
Optimal Objective Value: 205.70090440379596


From this new model, we're just purchasing 6 hotels: 
* Hotel 3
* Hotel 9
* Hotel 11
* Hotel 13
* Hotel 15
* Hotel 16

The total predicted profitability is now 205.7. As expected, this is a decrease from the previous model without the location constraint. 

#### **Question 6**

From the start, I would present the results of the analysis according to their requirement of diversification. I would emphasize that the selected hotels are Hotel 3, Hotel 9, Hotel 11, Hotel 13, Hotel 15, and Hotel 16. These hotels are strategically chosen to diversify the investment across different cities, aligning with LQM's preference for minimizing concentration risk. The optimal total predicted profitability is $205.7.

Regarding the regression model, I would recommend LQM to consider a few improvements. Firstly, they should conduct a thorough evaluation of the model's performance by analyzing various statistical measures such as R-squared, adjusted R-squared, and significance levels of coefficients. This analysis will provide insights into the overall goodness-of-fit and the importance of each independent variable in explaining profitability. Additionally, LQM could consider gathering more data points from a broader range of locations to improve the model's generalizability and accuracy. It is essential to ensure that the model is robust and reliable for predicting profitability in new locations.

As for the optimization model, I would suggest LQM explore incorporating additional factors such as hotel-specific characteristics, market conditions, or competitive landscape into the model. By expanding the set of independent variables, the optimization model can capture more nuanced factors that influence profitability.

### **Exercise 3**

#### **Question 1**

In [32]:
path_2 = "/OD.xlsx"

OD = pd.read_excel(path_2)

/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


In [36]:
OD.head()

Name  Gender  Sports  Theater  Religion  Outdoors  Eating Out
0     Laura  Female       0        2        -1         0           2
1  Jennifer  Female       0        2        -1        -1           1
2      Emma  Female       2       -1         2         0           2
3    Olivia  Female       0        2         0         1           1
4    Martha  Female       0       -1        -1         0           0

In [42]:
OD['Gender']

0     Female
1     Female
2     Female
3     Female
4     Female
5     Female
6     Female
7     Female
8       Male
9       Male
10      Male
11      Male
12      Male
13      Male
14      Male
15      Male
Name: Gender, dtype: object

In [46]:
import numpy as np
from scipy.optimize import linear_sum_assignment

# Extract the activity scores from the dataframe
activity_scores = OD[['Sports', 'Theater', 'Religion', 'Outdoors', 'Eating Out']].values

# Compute the weight matrix
weight_matrix = np.dot(activity_scores, activity_scores.T)

# Exclude self-matches by removing diagonal elements
np.fill_diagonal(weight_matrix, 0)

# Get the indices of male and female individuals
male_indices = OD[OD['Gender'] == 'Male'].index
female_indices = OD[OD['Gender'] == 'Female'].index

# Initialize the row and column indices for the linear sum assignment
row_indices = np.arange(len(male_indices))
col_indices = np.arange(len(female_indices))

# Apply the Hungarian algorithm to find the maximum weighted bipartite matching
row_indices, col_indices = linear_sum_assignment(-weight_matrix[male_indices][:, female_indices])

# Print the optimal matches and their aggregate match scores
aggregate_score = 0
matched = set()
for i in range(len(row_indices)):
    male = OD['Name'].iloc[male_indices[row_indices[i]]]
    female = OD['Name'].iloc[female_indices[col_indices[i]]]
    if male not in matched and female not in matched:
        match_score = weight_matrix[male_indices[row_indices[i]], female_indices[col_indices[i]]]
        aggregate_score += match_score
        matched.add(male)
        matched.add(female)
        print(f"Match: {female} & {male}, Match Score: {match_score}")

print(f"\nAggregate Match Score: {aggregate_score}")

Match: Mary & James, Match Score: 5
Match: Sophia & Robert, Match Score: 5
Match: Olivia & Peter, Match Score: 4
Match: Martha & Eric, Match Score: 4
Match: Emma & Daniel, Match Score: 8
Match: Jennifer & Adam, Match Score: 6
Match: Laura & Carl, Match Score: 8
Match: Isabella & Ralph, Match Score: 8

Aggregate Match Score: 48


#### **Question 2**

In [53]:
import numpy as np
from heapq import nlargest

# Extract the activity scores from the dataframe
activity_scores = OD[['Sports', 'Theater', 'Religion', 'Outdoors', 'Eating Out']].values

# Compute the weight matrix
weight_matrix = np.dot(activity_scores, activity_scores.T)

# Exclude self-matches by removing diagonal elements
np.fill_diagonal(weight_matrix, 0)

# Get the indices of male and female individuals
male_indices = OD[OD['Gender'] == 'Male'].index
female_indices = OD[OD['Gender'] == 'Female'].index

# Initialize the lists to store the top two matches for each individual
top_matches = [[] for _ in range(len(male_indices))]

# Iterate over male individuals and find their top two matches
for i in range(len(male_indices)):
    top_female_indices = nlargest(2, range(len(female_indices)), key=lambda x: weight_matrix[male_indices[i], female_indices[x]])
    for idx in top_female_indices:
        top_matches[i].append((idx, weight_matrix[male_indices[i], female_indices[idx]]))

# Print the top two matches for each male individual
for male_idx, matches in enumerate(top_matches):
    male_name = OD['Name'].iloc[male_indices[male_idx]]
    print(f"Male: {male_name}")
    for female_idx, match_score in matches:
        female_name = OD['Name'].iloc[female_indices[female_idx]]
        print(f"\tFemale: {female_name}, Match Score: {match_score}")


Male: James
	Female: Emma, Match Score: 6
	Female: Mary, Match Score: 5
Male: Robert
	Female: Emma, Match Score: 5
	Female: Sophia, Match Score: 5
Male: Peter
	Female: Isabella, Match Score: 5
	Female: Olivia, Match Score: 4
Male: Eric
	Female: Emma, Match Score: 4
	Female: Martha, Match Score: 4
Male: Daniel
	Female: Emma, Match Score: 8
	Female: Olivia, Match Score: 2
Male: Adam
	Female: Laura, Match Score: 6
	Female: Jennifer, Match Score: 6
Male: Carl
	Female: Laura, Match Score: 8
	Female: Olivia, Match Score: 7
Male: Ralph
	Female: Isabella, Match Score: 8
	Female: Olivia, Match Score: 5
